src/runner.py

```diff
@@ -88,596 +47,416 @@ class LxdInstanceConfig(TypedDict):
-        """Construct RunnerManager object for creating and managing runners.
+    @retry(tries=10, delay=1, logger=logger)
+    def _ensure_runner_profile(self) -> None:
+        """Ensure the runner profile is present on LXD.
 
-        Args:
-            path: GitHub repository path in the format '<org>/<repo>', or the GitHub organization
-                name.
-            token: GitHub personal access token to register runner to the repository or
-                organization.
-            app_name: An name for the set of runners.
-            reconcile_interval: Number of minutes between each reconciliation of runners.
+        Raises:
+            RunnerError: Unable to create the runner profile.
         """
-        http_proxy = os.environ.get("JUJU_CHARM_HTTP_PROXY", None)
-        https_proxy = os.environ.get("JUJU_CHARM_HTTPS_PROXY", None)
-        no_proxy = os.environ.get("JUJU_CHARM_NO_PROXY", None)
-        self.proxies = {}
-        if http_proxy:
-            self.proxies["http"] = http_proxy
-        if https_proxy:
-            self.proxies["https"] = https_proxy
-        if no_proxy:
-            self.proxies["no_proxy"] = no_proxy
-        self.session = requests.Session()
-        if self.proxies:
-            # setup proxy for requests
-            self.session.proxies.update(self.proxies)
-            # add proxy to fastcore which ghapi uses
-            proxy = urllib.request.ProxyHandler(self.proxies)
-            opener = urllib.request.build_opener(proxy)
-            fastcore.net._opener = opener
-        self._jinja = Environment(loader=FileSystemLoader("templates"), autoescape=True)
-        self._lxd = pylxd.Client()
-        self.path = path
-        self._api = GhApi(token=token)
-        self.app_name = app_name
-        self.reconcile_interval = reconcile_interval
-
-    @classmethod
-    def install_deps(cls) -> None:
-        """Install dependencies."""
-        # Binding for snap, apt, and lxd init commands are not available so subprocess.run used.
-        subprocess.run(["/usr/bin/snap", "install", "lxd"], check=True)  # nosec 603
-        subprocess.run(["/snap/bin/lxd", "init", "--auto"], check=True)  # nosec 603
-        subprocess.run(  # nosec B603
-            [
-                "/usr/bin/apt",
-                "install",
-                "-qy",
-                "cpu-checker",
-                "libvirt-clients",
-                "libvirt-daemon-driver-qemu",
-            ],
-            check=True,
-        )
-        cls.runner_bin_path.parent.mkdir(parents=True, exist_ok=True)
-
-    def get_latest_runner_bin_url(self) -> Optional[str]:
-        """Get the URL for the latest runner binary.
+        if not self._lxd.profiles.exists("runner"):
+            logger.info("Creating runner LXD profile")
+            profile_config = {
+                "security.nesting": "true",
+                "security.privileged": "true",
```

Comment:
Aren't nesting and privileged risky ? I can see some possible risks possibility with a nested instance (like bypassing potential apparmor restrictions, access to /proc and /sys)

Reply:
I am unsure if these are needed.
This profile setting was there before.
We are request security team for a code review (in addition to the report they already produced) and they may help us understand the implication of this more.

Reply:
Well if that's not needed I'd suggest to remove them or at least the "privileged"